src/irmin-pack/inode.ml

```diff
@@ -1414,13 +1444,42 @@ struct
   type key = Key.t
   type value = Inter.Val.t
 
+  exception Invalid_depth = Inter.Raw.Invalid_depth
+
   let mem t k = Pack.mem t k
+  let pp_value = Irmin.Type.pp Inter.Raw.t
+
+  let err_invalid_depth expected got v =
+    raise (Invalid_depth { expected; got; v })
+
+  let pp_invalid_depth ppf (expected, got, v) =
+    Fmt.pf ppf "Invalid depth: got %d, expecting %d (%a)" got expected pp_value
+      v
+
+  let check_depth v e g = if g <> e then err_invalid_depth e g v
+
+  (* call be recursive inners calls to find *)
+  let check_depth_inner ~init_depth ~depth v =
+    match v with
+    | None -> ()
+    | Some v -> (
+        match (init_depth, Inter.Raw.depth v) with
+        | None, _ ->
+            (* recursive calls only happens on non-value root nodes. *)
+            assert false
+        | Some d0, Some d -> check_depth v depth (d0 + d)
+        | Some _, None -> ())
 
   let find t k =
     Pack.find t k >|= function
     | None -> None
     | Some v ->
-        let find = Pack.unsafe_find ~check_integrity:true t in
+        let init_depth = Inter.Raw.depth v in
+        let find ~depth k =
+          let v = Pack.unsafe_find ~check_integrity:true t k in
+          check_depth_inner ~init_depth ~depth v;
+          v
+        in
         let v = Val.of_raw find v in
         Some v
```

Comment:
I think that `init_depth` is always `Node` or `Some 0`.  If I'm not mistaken, that find function is always called on root inodes.